tests/unit/libutil/strings.cc

```diff
@@ -231,4 +232,152 @@ TEST(tokenizeString, tokenizeSepEmpty)
+}
+
+TEST(splitString, tokenizeTabsSpacesNewlineWithDefaults)
+{
+    auto s = "foo\t\n bar\t\n baz";
+    // Using it like this is weird, but shows the difference with tokenizeString, which also has this test
+    Strings expected = {"foo", "", "", "bar", "", "", "baz"};
+
+    ASSERT_EQ(splitString<Strings>(s, " \t\n\r"), expected);
+}
+
+TEST(splitString, tokenizeTabsSpacesNewlineRetWithDefaults)
+{
+    auto s = "foo\t\n\r bar\t\n\r baz";
+    // Using it like this is weird, but shows the difference with tokenizeString, which also has this test
+    Strings expected = {"foo", "", "", "", "bar", "", "", "", "baz"};
+
+    ASSERT_EQ(splitString<Strings>(s, " \t\n\r"), expected);
+
+    auto s2 = "foo \t\n\r bar \t\n\r baz";
+    Strings expected2 = {"foo", "", "", "", "", "bar", "", "", "", "", "baz"};
+
+    ASSERT_EQ(splitString<Strings>(s2, " \t\n\r"), expected2);
+}
+
+TEST(splitString, tokenizeWithCustomSep)
+{
+    auto s = "foo\n,bar\n,baz\n";
+    Strings expected = {"foo\n", "bar\n", "baz\n"};
+
+    ASSERT_EQ(splitString<Strings>(s, ","), expected);
+}
+
+TEST(splitString, tokenizeSepAtStart)
+{
+    auto s = ",foo,bar,baz";
+    Strings expected = {"", "foo", "bar", "baz"};
+
+    ASSERT_EQ(splitString<Strings>(s, ","), expected);
+}
+
+TEST(splitString, tokenizeSepAtEnd)
+{
+    auto s = "foo,bar,baz,";
+    Strings expected = {"foo", "bar", "baz", ""};
+
+    ASSERT_EQ(splitString<Strings>(s, ","), expected);
+}
+
+TEST(splitString, tokenizeSepEmpty)
+{
+    auto s = "foo,,baz";
+    Strings expected = {"foo", "", "baz"};
+
+    ASSERT_EQ(splitString<Strings>(s, ","), expected);
+}
+
+// TODO: remove when concatStringsSep is fixed
+template<class C>
+std::string concatStringsSep2(const std::string_view sep, const C & ss)
+{
+    size_t size = 0;
+    bool tail = false;
+    // need a cast to string_view since this is also called with Symbols
+    for (const auto & s : ss) {
+        if (tail)
+            size += sep.size();
+        size += std::string_view(s).size();
+        tail = true;
+    }
+    std::string s;
+    s.reserve(size);
+    tail = false;
+    for (auto & i : ss) {
+        if (tail)
+            s += sep;
+        s += i;
+        tail = true;
+    }
+    return s;
+}
+
+// concatStringsSep sep . splitString sep = id   if sep is 1 char
+RC_GTEST_PROP(splitString, recoveredByConcatStringsSep, (const std::string & s))
+{
+    if (!s.empty()) {
+        // See FIXME in concatStringsSep tests
+        // RC_ASSERT(
+        //     concatStringsSep(std::string(1, s[0]), splitString<Strings>(s, std::string(1, s[0]))) == s
+        // );
+
+        // Work around the bug. Make sure it doesn't start with the separate, because that's broken.
+        RC_ASSERT(concatStringsSep2(std::string(1, s[0]), splitString<Strings>(s, std::string(1, s[0]))) == s);
+    }
+
+    RC_ASSERT(concatStringsSep2("/", splitString<Strings>(s, "/")) == s);
+    RC_ASSERT(concatStringsSep2("a", splitString<Strings>(s, "a")) == s);
```

Comment:
I believe this can be resolved now, since https://github.com/NixOS/nix/pull/11093

```suggestion
    RC_ASSERT(concatStringsSep("/", splitString<Strings>(s, "/")) == s);
    RC_ASSERT(concatStringsSep("a", splitString<Strings>(s, "a")) == s);
```